python/hyperspace/hyperspace.py

```diff
@@ -0,0 +1,174 @@
+        >>> hyperspace.indexes()
+        """
+        return DataFrame(self.hyperspace.indexes(), self.spark._wrapped)
+
+    def createIndex(self, dataFrame, indexConfig):
+        """
+        Creates index on the given dataframe using the given indexConfig.
+        :param dataFrame: dataFrame
+        :param indexConfig: indexConfig
+
+        >>> hyperspace = Hyperspace(spark)
+        >>> idxConfig = IndexConfig("indexName", ["c1"], ["c2","c3"])
+        >>> df = spark.read.parquet("./sample.parquet").toDF("c1", "c2", "c3")
+        >>> hyperspace.createIndex(df, indexConfig)
+        """
+        self.hyperspace.createIndex(dataFrame._jdf, self._getJavaIndexConfig(indexConfig))
+
+    def deleteIndex(self, indexName):
+        """
+        Soft deletes given index.
+        :param indexName: index name
+
+        >>> hyperspace = Hyperspace(spark)
+        >>> hyperspace.deleteIndex("indexname")
+        """
+        self.hyperspace.deleteIndex(indexName)
+
+    def restoreIndex(self, indexName):
+        """
+        Restores index with given index name.
+        :param indexName: index name
+
+        >>> hyperspace = Hyperspace(spark)
+        >>> hyperspace.restoreIndex("indexname")
+        """
+        self.hyperspace.restoreIndex(indexName)
+
+    def vacuumIndex(self, indexName):
+        """
+        Vacuums index with given index name.
+        :param indexName: index name
+
+        >>> hyperspace = Hyperspace(spark)
+        >>> hyperspace.vacuumIndex("indexname")
+        """
+        self.hyperspace.vacuumIndex(indexName)
+
+    def refreshIndex(self, indexName):
+        """
+        Update indexes for the latest version of the data.
+        :param indexName: index name
+
+        >>> hyperspace = Hyperspace(spark)
+        >>> hyperspace.refreshIndex("indexname")
+        """
+        self.hyperspace.refreshIndex(indexName)
+
+    def cancel(self, indexName):
+        """
+        Cancel api to bring back index from an inconsistent state to the last known stable state.
+        :param indexName: index name
+
+        >>> hyperspace = Hyperspace(spark)
+        >>> hyperspace.cancel("indexname")
+        """
+        self.hyperspace.cancel(indexName)
+
+    def explain(self, df, verbose=False, redirectFunc=lambda x: sys.stdout.write(x)):
+        """
+        Explains how indexes will be applied to the given dataframe.
+        :param df: dataFrame
+        :param redirectFunc: optional function to redirect output of explain
+
+        >>> hyperspace = Hyperspace(spark)
+        >>> df = spark.read.parquet("./sample.parquet").toDF("c1", "c2", "c3")
+        >>> hyperspace.explain(df)
+        """
+        analyzer = self.jvm.com.microsoft.hyperspace.index.plananalysis.PlanAnalyzer
+        result_string = analyzer.explainString(df._jdf, self.spark._jsparkSession,
+                                               self.hyperspace.indexes(), verbose)
+        redirectFunc(self.jvm.java.lang.String(result_string))
```

Comment:
so `result_string` is not a Python string?

Reply:
removed cast. thanks